DOCSP-43081: Change Streams (#37)
mongodb/docs-c: source/read/change-streams.txt

.. _c-change-streams:

====================
Monitor Data Changes
====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: watch, code example

Overview
--------

In this guide, you can learn how to use the {+driver-short+} to monitor a  **change stream**,
allowing you to view real-time changes to your database. A change stream is a {+mdb-server+} feature that
publishes data changes on a collection, database, or deployment. Your application can
subscribe to a change stream and use events to perform other actions.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

Open a Change Stream
--------------------

To open a change stream, call one of the following functions that corresponds to the

> scope of events you want to observe.

Reviewer: ```suggestion
scope of events you want to observe:
```

> - ``mongoc_client_watch()``: To monitor all changes in the MongoDB deployment
> - ``mongoc_database_watch()``: To monitor changes in all collections in the database
> - ``mongoc_collection_watch()``: To monitor changes in the collection

Reviewer: ```suggestion
- ``mongoc_client_watch()``: Monitors all changes in the MongoDB deployment
- ``mongoc_database_watch()``: Monitors changes in all collections in the database
- ``mongoc_collection_watch()``: Monitors changes in the collection
```

application or shell, perform a write operation on the ``restaurants`` collection. The
following example updates a document in which the value of the ``name`` field is ``"Blarney Castle"``:

.. _c-change-stream-update:

.. literalinclude:: /includes/read/change-streams.c
   :start-after: start-update-for-change-stream
   :end-before: end-update-for-change-stream
   :language: c
   :copyable:
   :dedent:

When you update the collection, the change stream application prints the change
as it occurs. The printed change event resembles the
following:

.. code-block:: json

   {
     "_id": { ... },
     "operationType": "update",
     "clusterTime": { ... },
     "ns": {
       "db": "sample_restaurants",
       "coll": "restaurants"
     },
     "updateDescription": {
       "updatedFields": {
         "cuisine": "Irish"
       },
       "removedFields": [],
       "truncatedArrays": []
     }
     ...
   }

Modify the Change Stream Output
-------------------------------

You can pass the ``pipeline`` parameter to any watch function to modify the
change stream output. This parameter allows you to watch for only specified

> change events. Format the parameter as a list of object, where each object represents an

Reviewer: ```suggestion
change events. Format the parameter as a list of objects, where each object represents an
```

> change events. Format the parameter as a list of objects that each represents an
> aggregation stage.

Reviewer: S:
```suggestion
change events. Format the parameter as a list of objects, where each object represents an
aggregation stage.
```

You can specify the following stages in the ``pipeline`` parameter:

- ``$addFields``
- ``$match``
- ``$project``
- ``$replaceRoot``
- ``$replaceWith``
- ``$redact``
- ``$set``
- ``$unset``

> Match Specific events Example

Reviewer: ```suggestion
Match Specific Events Example
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example uses the ``pipeline`` parameter to include a ``$match`` stage

> to open a change stream that only records update operations:

Reviewer: ```suggestion
to open a change stream that records only update operations:
```
